### PR TITLE
Collect `gardener_operator_garden_condition` metric

### PR DESCRIPTION
**Pull request by maboehm**
**How to categorize this PR?**
/area monitoring
/kind enhancement

**What this PR does / why we need it**:
Start collecting metrics about resources managed by the gardener-operator.

Currently, `gardener/gardener` barely exports any custom resource metrics, since that is handled by the [gardener-metrics-exporter](https://github.com/gardener/gardener-metrics-exporter). This component, however, only watches resources in the `virtualGarden` cluster, but the resources managed by the operator are not in that cluster. I therefore propose to start exporting the metrics here.

I started with only one condition, `gardener_operator_garden_condition` which is similar to the `shoot_condition` metric [here](https://github.com/gardener/gardener-metrics-exporter/blob/master/pkg/metrics/metrics.go#L106) (just with fewer labels)

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
`gardener-operator` now collects the metric `gardener_operator_garden_condition`
```

### CODE REVIEW

File: pkg/operator/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		gardenCollector: newGardenCollector(k8sClient, mgr.GetLogger().WithName("operator-metrics")),
+	})
+}
+
+func (r *runnable) Start(_ context.Context) error {
+	runtimemetrics.Registry.MustRegister(r.gardenCollector)
+	return nil
+}
+
+func mapConditionStatus(status gardencorev1beta1.ConditionStatus) float64 {
```
Comment (timebertt) [inline pkg/operator/metrics/metrics.go:39]: I suggest not to map condition status to metric values. Instead, let's use a stateset-style metric like kube-state-metrics does. I.e., there should be a `status` label for all possible states and the current status series should have the value `1`, while others are `0`.

File: pkg/operator/metrics/garden.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		c.log.Error(err, "Failed to list gardens")
+		return
+	}
+
+	for i := range gardenList.Items {
+		garden := gardenList.Items[i]
+		if garden.Status.LastOperation == nil {
+			continue
+		}
+		lastOperation := string(garden.Status.LastOperation.Type)
```
Comment (timebertt) [inline pkg/operator/metrics/garden.go:66]: The last operation should probably be exported in a dedicated metric. Otherwise, it results in a high cardinality of the condition metric.
I know this won't be a problem for gardener-operator, where we only have a single `Garden` object at a time. However, we should follow best practices here.

File: pkg/operator/metrics/garden.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+		for i := range garden.Status.Conditions {
+			condition := garden.Status.Conditions[i]
+			if condition.Type == "" {
+				continue
+			}
```
Comment (timebertt) [inline pkg/operator/metrics/garden.go:72]: The condition variable doesn't escape the loop iteration, so you could use `for i, condition := range ...`.

**Review by timebertt — COMMENTED** [review]
Nice PR!
/ok-to-test

**Comment by rickardsjp** [comment]
Hi @maboehm, thanks for the PR! Exposing the `Garden`-resource's conditions is something we've been looking into as well: So far, metrics about custom Gardener resources have been exposed by [gardener-metrics-exporter](https://github.com/gardener/gardener-metrics-exporter). However, the `Garden` resource has been the exception because unlike Shoots, Seeds, Projects, etc., its a custom resource in the runtime cluster. Not exposing any metrics about this resource has been a bit of a thorn in our eye as well. I appreciate all the effort that went into this PR, yet I want to present an alternative:

With #10062, the garden-runtime cluster now contains a kube-state-metrics deployment. #9941 upgrades the version of KSM used in Gardener to one that supports [custom resource state metrics](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md). Once #9941 is merged, exposing metrics on the `Garden` resource's conditions would be as easy as extending the CSR configuration with something like the following snippet:

```
  - groupVersionKind:
      group: operator.gardener.cloud
      kind: "Garden"
      version: "v1alpha1"
    metricNamePrefix: "garden"
    labelsFromPath:
      name:
      - metadata
      - name
    metrics:
    - name: "garden_condition"
      help: "represents a condition of a Garden object"
      each:
        type: StateSet
        stateSet:
          labelName: condition
          labelsFromPath:
            type: [type]
          path: [status, conditions]
          valueFrom: ["status"]
          list: [Progressing, True, False, Unknown]
```

That comes at the cost of some flexibility but also spares us a bunch of code. What do you think?

**Comment by maboehm** [comment]
@rickardsjp 
I was not aware that this effort is going on. It sounds really interesting, but was a bit hard to follow the overall picture (including the end-goal of maybe getting rid of the metrics-exporter), without an umbrella issue.

I think I also prefer this approach, and would wait for https://github.com/gardener/gardener/pull/9941 to be merged, and then I can add a CR-metric for the garden condition & operation for the `runtime` KSM.

**Comment by rickardsjp** [comment]
>I was not aware that this effort is going on. It sounds really interesting, but was a bit hard to follow the overall picture (including the end-goal of maybe getting rid of the metrics-exporter), without an umbrella issue.

100% agreed, that's on us. Because the garden-level monitoring only moved into the open-source relatively recently, I think we (the SAP-side Gardener monitoring contributors) have not been very good at ensuring we communicate that transparently. There are also still some doubt whether removing the gardener-metrics-exporter entirely is feasible. Nevertheless, we should have made this effort easier to identify with an umbrella issue.

>add a CR-metric for the garden condition & operation for the runtime KSM.

I think the last operation is also valid as long as it is it's own metric. It can only have one of [five values](https://github.com/gardener/gardener/blob/master/pkg/apis/core/v1beta1/types_common.go#L64-L75), so there are no cardinality concerns. Building on the snippet I pasted above, I just quickly added this:

```
    - name: "garden_last_operation"
      help: "denotes the last operation performed on a Garden object"
      each:
        type: StateSet
        stateSet:
          labelName: last_operation
          path: [status, lastOperation]
          valueFrom: ["type"]
          list: [Create, Reconcile, Delete, Migrate, Restore]
```

which produces this output:

```
# HELP garden_garden_last_operation denotes the last operation performed on a Garden object
# TYPE garden_garden_last_operation stateset
garden_garden_last_operation{customresource_group="operator.gardener.cloud",customresource_kind="Garden",customresource_version="v1alpha1",last_operation="Create",name="garden"} 0
garden_garden_last_operation{customresource_group="operator.gardener.cloud",customresource_kind="Garden",customresource_version="v1alpha1",last_operation="Delete",name="garden"} 0
garden_garden_last_operation{customresource_group="operator.gardener.cloud",customresource_kind="Garden",customresource_version="v1alpha1",last_operation="Migrate",name="garden"} 0
garden_garden_last_operation{customresource_group="operator.gardener.cloud",customresource_kind="Garden",customresource_version="v1alpha1",last_operation="Reconcile",name="garden"} 1
garden_garden_last_operation{customresource_group="operator.gardener.cloud",customresource_kind="Garden",customresource_version="v1alpha1",last_operation="Restore",name="garden"} 0
```

Once 9941 is through, I will add both snippets to the custom resource state ConfigMap, adapt the `kube-state-metrics` scrape config to keep the Garden metrics, and we should have Garden resource metrics. 🎉
